Docs: Show primary story description and headline in autodocs

tmeasday opened the pull request:
Issue: N/A

Telescoping on https://github.com/storybookjs/storybook/pull/20559

## What I did

Added a render option `forceInitialArgs` which means a story is rendered with it's initial args no matter what and doesn't respond to arg updates. Use this in the `<Stories/>` block, and switch to showing the primary story there.

The idea is that now in Autodocs we show:

- Primary story, reactive
- Controls (to control ☝️)
- All stories (inc. primary) with descriptions, unreactive.

![image](https://user-images.githubusercontent.com/132554/212579681-fe3a0a05-b213-40ad-b361-bc37cf0db641.png)


## How to test

- Look at tests + stories
- Try autodocs!

JReinhold commented on code/lib/preview-api/src/modules/preview-web/PreviewWeb.test.ts:
```diff
@@ -1131,6 +1131,39 @@ describe('PreviewWeb', () => {
           );
         });
+
+        it('does not re-renders the story when ignoreArgsUpdates=true', async () => {
```
```suggestion
        it('does not re-render the story when ignoreArgsUpdates=true', async () => {
```

JReinhold commented on code/ui/blocks/src/components/Story.stories.tsx:
```diff
@@ -48,9 +48,38 @@ export const IFrame = {
+    ignoreArgsUpdates: true,
+    renderStoryToElement,
+  },
+  play: async ({ args, canvasElement, loaded }: PlayFunctionContext<WebRenderer>) => {
```
```suggestion
  // test that it ignores updated args by emitting an arg update and assert that it isn't reflected in the DOM
  play: async ({ args, canvasElement, loaded }: PlayFunctionContext<WebRenderer>) => {
```

JReinhold reviewed (approved):
Looks good! 👍

tmeasday committed:
Change to `forceInitialArgs`

tmeasday committed:
Suggestion

tmeasday commented:
Ok, now this is ready for re-review. Sorry for the early review @JReinhold. I changed the name and added a test for the "initial" behaviour.

I wasn't sure if we wanted to add stories for the `Stories` and `DocsStory` components. I sort of feel like they'll be duplicative and it'll be hard to actually test it's working outside of more copies of the same play function.

tmeasday committed:
Merge branch 'tom/sb-1144-update-api-of-story-block' into add-ignoreArgsUpdates

tmeasday committed:
Merge branch 'tom/sb-1144-update-api-of-story-block' into add-ignoreArgsUpdates

tmeasday committed:
Fix Preview stories

tmeasday committed:
Fix ordering spec

tmeasday commented:
@JReinhold so the e2e test is failing due to this query:

https://github.com/storybookjs/storybook/blob/c22aac8e5bb8a85bb81fd58db1d9664a9662e6c3/code/e2e-tests/addon-docs.spec.ts#L49

I guess that's because the `Story` component generates a DOM id based on the story's id: 

https://github.com/storybookjs/storybook/blob/c22aac8e5bb8a85bb81fd58db1d9664a9662e6c3/code/ui/blocks/src/components/Story.tsx#L60

What shall we do about this do you think?

shilman commented:
I think the primary story should have a special ID. not sure what though. Not sure what though `<id>--primary`?--or how it would impact anything

tmeasday committed:
Add a `__primary` prop and use it to alter the idea of `<Primary>`

tmeasday commented:
Ok, that's make sense. I did that.

tmeasday committed:
Fix up some props

tmeasday committed:
Update test to cope with story rendering more than once

tmeasday commented:
@valentinpalkovic the failure on this branch comes from the angular renderer which doesn't appear to be able to render the same story more than once on the screen:

![image](https://user-images.githubusercontent.com/132554/212595803-90492568-09ac-4b35-a903-8650e4e953f3.png)

Is that surprising to you? If it's expected would it be easy to change/fix?

tmeasday committed:
Merge branch 'tom/sb-1144-update-api-of-story-block' into add-ignoreArgsUpdates

valentinpalkovic commented:
Fixed in https://github.com/storybookjs/storybook/pull/20559 :)

tmeasday commented:
Oh amazing!

JReinhold commented:
@valentinpalkovic 

> Fixed in https://github.com/storybookjs/storybook/pull/20559 :)

@tmeasday

Are you sure about that Valentin? The issue you and I discussed a few days ago, was about Angular being unable to render any multiple stories twice, like it does in docs. But this here is specifically about rendering _the same_ story twice.

Maybe you've already taken care of this since our last conversation, then please ignore me.

JReinhold commented on code/ui/blocks/src/blocks/Story.stories.tsx:
```diff
@@ -199,3 +200,12 @@ export const WithInteractionsAutoplayInStory: Story = {
   },
 };
+
+export const IgnoreArgsUpdates: Story = {
```
```suggestion
export const ForceInitialArgs: Story = {
```

JReinhold reviewed (approved):
LGTM! 👍

valentinpalkovic commented:
> @valentinpalkovic
> 
> > Fixed in #20559 :)
> 
> @tmeasday
> 
> Are you sure about that Valentin? The issue you and I discussed a few days ago, was about Angular being unable to render any multiple stories twice, like it does in docs. But this here is specifically about rendering _the same_ story twice.
> 
> Maybe you've already taken care of this since our last conversation, then please ignore me.

Angular components are rendered as Standalone components in https://github.com/storybookjs/storybook/pulls. Therefore, the same component can be rendered multiple times without any issues.

tmeasday committed:
Update code/ui/blocks/src/blocks/Story.stories.tsx

Co-authored-by: Jeppe Reinhold <jeppe@chromatic.com>

tmeasday committed:
Merge remote-tracking branch 'origin/valentin/angular15-maintenance' into add-ignoreArgsUpdates

tmeasday commented:
I changed the base of this https://github.com/storybookjs/storybook/pull/20559, let's find out for sure!

valentinpalkovic commented:
@tmeasday Hmm. Still, it does not work. let me take a quick look. Shouldn't be that hard to fix.

valentinpalkovic committed:
Fixed Angular components on Docs paged rendered Story twice

tmeasday committed:
Fix typing

tmeasday committed:
Still trying to fix typing

tmeasday committed:
Merge remote-tracking branch 'origin/next' into add-ignoreArgsUpdates

chakAs3 commented:
In my opinion, The Main Story, Default or Playground should be generated as the Base Story from the component definition only ( with default values ) and be independent of all Stories. 
Base/Default Story is the Interactive Playground that can match the state of all the stories and not only the first one.

I used the same technique in most Storybook projects 

![image](https://user-images.githubusercontent.com/711292/213696369-fffd8922-b1e8-4ff8-afcf-eb6ec0df50a8.png)

so the Playground on where you mix and match different states of your component, and mostly the following stories are complete illustrations of Variants so the user can see a max of the component's states. 

i have been raising some issues since the update.

JReinhold commented:
> In my opinion, The Main Story, Default or Playground should be generated as the Base Story from the component definition only ( with default values ) and be independent of all Stories. Base/Default Story is the Interactive Playground that can match the state of all the stories and not only the first one.
> 
> I used the same technique in most Storybook projects
> 
> ![image](https://user-images.githubusercontent.com/711292/213696369-fffd8922-b1e8-4ff8-afcf-eb6ec0df50a8.png)
> 
> so the Playground on where you mix and match different states of your component, and mostly the following stories are complete illustrations of Variants so the user can see a max of the component's states.
> 
> i have been raising some issues since the update.

I don't disagree with you, but in the current data model I don't think that's possible. There's no such thing as displaying a story only based on the meta information, only actual stories can be displayed.

And what would you expect to be shown with the following stories file?
```tsx
export default {
  title: 'My Button'
}

export const Primary = {
  render: () => <button>hello</button>
}
```

Without a `component` or a `render` function on meta, we can't render anything solely based on the meta information.

chakAs3 commented:
> > In my opinion, The Main Story, Default or Playground should be generated as the Base Story from the component definition only ( with default values ) and be independent of all Stories. Base/Default Story is the Interactive Playground that can match the state of all the stories and not only the first one.
> > I used the same technique in most Storybook projects
> > ![image](https://user-images.githubusercontent.com/711292/213696369-fffd8922-b1e8-4ff8-afcf-eb6ec0df50a8.png)
> > so the Playground on where you mix and match different states of your component, and mostly the following stories are complete illustrations of Variants so the user can see a max of the component's states.
> > i have been raising some issues since the update.
> 
> I don't disagree with you, but in the current data model I don't think that's possible. There's no such thing as displaying a story only based on the meta information, only actual stories can be displayed.
> 
> And what would you expect to be shown with the following stories file?
> 
> ```tsx
> export default {
>   title: 'My Button'
> }
> 
> export const Primary = {
>   render: () => <button>hello</button>
> }
> ```
> 
> Without a `component` or a `render` function on meta, we can't render anything solely based on the meta information.

I know that's why i said generated Story we generate what we need to render the Story  based on the Component + meta info 
I need a component and args to render and if there is a Story context that will be great more info, 

from your code, I expect a  hello button

GuillaumeNury commented:
@tmeasday hello! What is the motivation of changing the default value of `includePrimary` ?
I cannot find a way to change this value.

JReinhold commented:
> @tmeasday hello! What is the motivation of changing the default value of `includePrimary` ? I cannot find a way to change this value.

1. It seemed less magical and more obvious to us that the list of Stories shows all stories, and didn't "hide" anything.
2. Our new support for story-level descriptions (along side component-level descriptions) needed a place to show the description for the primary story, and showing it in the list with the rest of the stories made this a non-issue.

I believe you can get back the old behavior by providing your own custom `DocsPage` that largely copies the existing one:

https://storybook.js.org/docs/7.0/react/writing-docs/docs-page#write-a-custom-template

```ts
// .storybook/preview.js

import React from 'react';
import { Title } from './Title';
import { Subtitle } from './Subtitle';
import { Description } from './Description';
import { Primary } from './Primary';
import { Controls } from './Controls';
import { Stories } from './Stories';

export const parameters = {
  ...
  docs: {
    page: () => (
	  <>
	    <Title />
	    <Subtitle />
	    <Description />
	    <Primary />
	    <Controls />
	    <Stories includePrimary={false} />
	  </>
	);
  },
};
```

GuillaumeNury commented:
Ok. We use Angular, I can see that I should dig around `mdx` file ([doc here](https://storybook.js.org/docs/7.0/angular/writing-docs/docs-page#write-a-custom-template)).
Our design system is documented using [ZeroHeight](https://zeroheight.com/). We display only one story with its controls on each page ([exemple here](https://prisme.lucca.io/94310e217/p/07849c-radio) which exposes [this story](https://lucca-front.lucca.io/storybook/?path=/docs/documentation-forms-radios-basic--docs)). Currently, we hide additional stories using CSS 🙈

GuillaumeNury commented:
@JReinhold I played with custom Docs template and it worked like a charm. Thank you!
